### DISCUSSION TITLE
Enabling UART and Network interface on boot

### DISCUSSION THREAD

**Discussion by Hariom-avin** [OE4T/meta-tegra#886] [Q&A] [answered]
Hello,

I am using NVIDIA Xavier hardware and using doflash for flashing the image.

I am using meta-tegra for the build.

I am not able to enable the uart serial terminal and Ethernet interface on boot. To enable the Ethernet interface i need to give command to make the interface up.

I want this Ethernet interface to get enabled as soon as image boots up.

I request you all to please suggest me how I can do it.

Regards,
Hariom

**Comment by madisongh** [comment]
Can you be more specific about what you're doing and what the problem is?

- Which Xavier model is this, and is it a development kit, or custom hardware of some kind?
- What image are you building?
- What do you mean by the UART and the Ethernet interface being "not enabled"?
- Which UART?

**Comment by Hariom-avin** [comment, marked as answer]
Hi Madisongh,

I am verifying Yocto build on NVIDIA hardware and I want to observe the logs on booting the flashed image.

1. It is NVIDIA JETSON XAVIER hardware.
2. I am building core-image-minimal image.
3. My intention was to say that, i am not able to see the terminal output using UART as well i am not able to ping the hardware. Also if i am trying to get the interface details using ifconfig command i am getting blank output. I also verified the /etc/network/interfaces file it is having a valid IP configuration inside it.
To make the network interface up user need to manually make it up by giving ifup command.
4. I want to see serial terminal output using micro USB which works as UART for terminal logs.

Regards,
Hariom

**Reply by madisongh** [reply]
As I said, there are multiple Xavier models.  It sounds like you're using a Jetson AGX Xavier development kit (as opposed to a Jetson Xavier NX development kit), but I was asking to make sure.

core-image-minimal is *minimal* - that is, it includes just enough to boot the device.  You'll probably need to include more packages in your image to automatically start the network.  We do have a [demo distro](https://github.com/OE4T/tegra-demo-distro) setup you can use for building some images to test with.

You're using the USB micro-B connector to access the serial console, and you say you're not seeing *any* console output?  Nothing from the bootloaders or Linux?  Which `/dev/ttyUSB<n>` device on your host system are you trying to use (assuming you're using Linux on your host), and do you have it configured properly (115200, 8, n)?  It should be `/dev/ttyUSB0` for the AGX Xavier.

**Reply by Hariom-avin** [reply]
Thanks madisong for reply.

**Reply by Hariom-avin** [reply]
Hi madisongh,

I am able to make network interface up. But now i am facing new issue. I request you to please support me.

Actually one of my application use OpenCV. And in build i am getting following error.

|   Could NOT find CUDA (missing: CUDA_TOOLKIT_ROOT_DIR CUDA_NVCC_EXECUTABLE
|   CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) (Required is exact version "10.2")

I have following settings in local.conf 
------------------------------------------------------------------------------------------
MACHINE ?= "jetson-agx-xavier-devkit"

DISTRO ?= "poky"
#DISTRO_FEATURES += "x11 opengl"

IMAGE_CLASSES += "image_types_tegra"
IMAGE_FSTYPES = "tegraflash"

SSTATE_DIR ?= "/home/${USER}/yocto/sstate_dir"
DL_DIR ?= "/home/${USER}/yocto/downloads"

PREFERRED_VERSION_python3 = "3.6%"


BB_NUMBER_THREADS = '2'
PARALLEL_MAKE = '-j2'

NVIDIA_DEVNET_MIRROR="file:///home/${USER}/Downloads/nvidia/sdkm_downloads" 

IMAGE_INSTALL_append = " cuda-samples"
EXTRA_IMAGE_FEATURES += " tools-sdk"

#GCCVERSION = "8.%"

GCCVERSION = "7.%"
require contrib/conf/include/gcc-compat.conf
--------------------------------------------------------------------------------------------------
And i have downloaded and installed packages using Nvidia sdk manager. 

I request you to please help me in resolving this issue.

Regards,
Hariom

**Reply by madisongh** [reply]
Which branch are you using?  You should not need to use NVIDIA_DEVNET_MIRROR with recent (dunfell and later) branches.

Are you trying to build your application on the device, or do you have a recipe for it in your Yocto setup? Does your application use CUDA?

**Reply by Hariom-avin** [reply]
Hi Madisongh,

I am using dunfell branch.

I want to build my application on the host machine using Yocto. Yes I have a recipe for it. Application don't have direct dependency on CUDA, but it have dependency on OpenCv.

**Reply by madisongh** [reply]
OK, well, we build OpenCV with CUDA support enabled by default, and it looks like OpenCV's Find module for CMake looks for CUDA when that's the case. Adding `inherit cuda` to the recipe for your application should resolve the issue.

**Reply by Hariom-avin** [reply]
Hi Madisongh, 

Thanks for your reply. I was able to compile my application. Also i was able to flash it on the target board. 

Here i faced one issue more issue and it was with respect to ssh login. I was using 'root' as user and 'root' as password. But i was getting "Permission denied, please try again." message. 

So my doubt is what login details i should use to login with ssh connection?

Regards,
Hariom

**Reply by madisongh** [reply]
For development, it's usually easiest to configure your image to allow root login with no password, using the `empty-root-password` and `allow-empty-password` settings in IMAGE_FEATURES. There's more info about this in [the Yocto Project documentation](https://docs.yoctoproject.org/singleindex.html#image-features).

**Reply by Hariom-avin** [reply]
Hi Madisong, 

Thanks for reply.

I tried above solution and I am able to login the terminal without password. But when i am trying to login from ssh I am getting the "Permission denied" message. I am using ssh root@192.168.7.10 as the command.

Following are the changes done in local.conf fie:
EXTRA_IMAGE_FEATURES += "allow-empty-password"
EXTRA_IMAGE_FEATURES += "empty-root-password"

**Reply by madisongh** [reply]
I think you need `allow-root-login` as well

**Reply by Hariom-avin** [reply]
Hi Madisongh,

Thanks for your reply, I checked it and it is working fine for me.

I have one more question Madisongh. I want to share NVIDIA Xavier image present in my build folder with one of my colleague, then my college will be flashing this image to the target hardware.
So my doubt is, what are the various files i need to share so it can be flashed on target board using 'doflash' script.

Regards,
Hariom

**Reply by madisongh** [reply]
They only need the [tegraflash package](https://github.com/OE4T/meta-tegra/wiki/Flashing-the-Jetson-Dev-Kit#building-a-tegraflash-package). It contains everything needed for flashing. There are other notes on that linked wiki page that you may want to share about the flashing process, too.

**Reply by Hariom-avin** [reply]
Hi Madisongh,

Thanks for your support. I was able to solve all technical difficulties.  Thanks again.
